### PR TITLE
ci: update build script, integrate with circleci 

### PR DESCRIPTION
**Pull request by gszr**
### Summary

- Update build script, automating builds for all distributions and architectures
- Integrate build script with CircleCI

Releases are automatically built and pushed to Bintray whenever a tag is pushed.

### CODE REVIEW

File: tools/distros.sh [hunk 1]
```diff
@@ -2,12 +2,133 @@
...
-for os in "${GOOS[@]}"
-do
-    for arch in "${GOARCH[@]}"
-    do
-        make GOOS=$os GOARCH=$arch build/artifact-tarball
-        curl -T build/artifacts-$os-$arch/kuma-$os-$arch.tar.gz -u $BINTRAY_USERNAME:$BINTRAY_API_KEY "https://api.bintray.com/content/kong/kuma/$os/$RELEASE_TAG-$arch/kuma-$RELEASE_TAG-$os-$arch.tar.gz?publish=1&override=1"
+
+BINTRAY_ENDPOINT="https://api.bintray.com/"
+BINTRAY_SUBJECT="kong"
+BINTRAY_REPOSITORY="kuma-tests"
```
Comment (gszr) [inline tools/distros.sh]: To be changed.

File: tools/releases/distros.sh [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  cp -p build/envoy-$distro $dest_dir/bin/envoy
+  cp -p build/artifacts-$system-$arch/kuma-cp/kuma-cp $dest_dir/bin
+  cp -p build/artifacts-$system-$arch/kuma-dp/kuma-dp $dest_dir/bin
+  cp -p build/artifacts-$system-$arch/kumactl/kumactl $dest_dir/bin
+  cp -p build/artifacts-$system-$arch/kuma-tcp-echo/kuma-tcp-echo $dest_dir/bin
+  cp -p pkg/config/app/kuma-cp/kuma-cp.defaults.yaml $dest_dir/conf/kuma-cp.conf
+
+  cp tools/releases/templates/* $dest_dir
+
+  tar -czf build/artifacts-$system-$arch/kuma-$system-$arch.tar.gz -C $dest_dir .
```
Comment (yskopets) [inline tools/releases/distros.sh]: when I run `./tools/releases/distros.sh --package --version 1.2.3` on my machine, I get only 2 `*.tar.gz` files in the end:

* `build/artifacts-darwin-amd64/kuma-darwin-amd64.tar.gz`
* `build/artifacts-linux-amd64/kuma-linux-amd64.tar.gz`

is it an intended behaviour ?

I would expect to have

* `kuma-*-ubuntu.tar.gz`
* `kuma-*-debian.tar.gz`
* `kuma-*-rhel.tar.gz`
* `kuma-*-centos.tar.gz`
* `kuma-*-darwin.tar.gz`

File: tools/releases/distros.sh [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+
+function release {
+  for os in "${DISTRIBUTIONS[@]}"; do
+    local distro=$(echo "$os" | awk '{split($0,parts,":"); print parts[1]}')
+    local system=$(echo "$os" | awk '{split($0,parts,":"); print parts[2]}')
+
+    for arch in "${GOARCH[@]}"; do
+      local artifact="build/artifacts-$system-$arch/kuma-$system-$arch.tar.gz"
```
Comment (yskopets) [inline tools/releases/distros.sh]: notice that artifact name doesn't include `distro`, which means that `centos` and `ubuntu` will be represented by the same `*.tar.gz` archive (which is not correct - `envoy` binary must differ)

File: .circleci/config.yml [hunk 4]
```diff
@@ -329,13 +329,19 @@ jobs:
...
     steps:
     - checkout
+    - restore_cache:
+        keys:
+        - go.mod/{{ checksum "go.sum" }}
     - run:
-        name: Push Docker images
-        command: make images/push BINTRAY_USERNAME=$BINTRAY_USERNAME BINTRAY_API_KEY=$BINTRAY_API_KEY KUMA_VERSION=$CIRCLE_TAG
+        name: Build packages
+        command: ./tools/releases/distros.sh --package --version $CIRCLE_TAG
```
[the remark this answers is not in the document]
Comment (gszr) [reply .circleci/config.yml]: @yskopets The `--sha` (optional) flag was introduced for that - you need to pass it.

File: tools/releases/distros.sh [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        usage
+        break
+        ;;
+    esac
+    shift
+  done
+
+  [ -z "$BINTRAY_USERNAME" ] && msg_err "BINTRAY_USERNAME required"
+  [ -z "$BINTRAY_API_KEY" ] && msg_err "BINTRAY_API_KEY required"
+  [ -z "$KUMA_VERSION" ] && msg_err "Error: --version required"
```
[the remark this answers is not in the document]
Comment (gszr) [reply tools/releases/distros.sh]: I'd prefer not to; it'd be unnecessary burden for local non-release usage. It is called in CI, though (just saw it wasn't, fixed it).

File: tools/releases/distros.sh [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  [ "$status" -ne "200" ] && msg_err "Error: failed downloading Envoy"
+}
+
+
+function create_tarball {
+  local system=$1
+  local arch=$2
+  local distro=$3
+
+  local dest_dir=build/kuma-$system-$arch
```
[the remark this answers is not in the document]
Comment (gszr) [reply tools/releases/distros.sh]: Binaries (with exception of Envoy) are system-specific -- not distro. But agreed, it'd make it harder to troubleshoot. Updated it!

File: tools/releases/distros.sh (hunk 2 above)
Comment (gszr) [reply tools/releases/distros.sh]: This must no longer be the case. It was, before we included Envoy in the tarballs.

File: tools/releases/distros.sh (hunk 3 above)
Comment (gszr) [reply tools/releases/distros.sh]: Fixed!

**Comment by yskopets** [comment]
@gszr awesome shell script, by the way.

Thanks a lot!

**Pull request merged**